Python tools
===============


## Sequence Analysis

### libraries

- biopython

    The Biopython Project is an international association of developers of freely available Python (https://www.python.org) tools for computational molecular biology. The Biopython web site (http://www.biopython.org) provides an online resource for modules, scripts, and web links for developers of Python-based software for bioinformatics use and research. Basically, the goal of Biopython is to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes. Biopython features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.
 
    [biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html)
    
### Online sources:

- websites

- books

- videos
    Presented for the course COMP 364 at McGill University:
 
    - 
        - [](https://www.youtube.com/watch?v=qQ7rIpB4oOw)
    In order to get the material of this video go to https://nbviewer.jupyter.org/github/cgoliver/Notebooks/blob/master/COMP_364/L25/L25.ipynb
    

## Processing Next Generation Sequencing (NGS) data

### libraries

- Biopython
- numpy
- seaborn

- [pysam](https://pysam.readthedocs.io/en/latest/)
    
    Pysam is a python module for reading, manipulating and writing genomic data sets.
    
    Pysam is a wrapper of the htslib C-API and provides facilities to read and write SAM/BAM/VCF/BCF/BED/GFF/GTF/FASTA/FASTQ files as well as access to the command line functionality of the samtools and bcftools packages. The module supports compression and random access through indexing.
    
- [GATK](https://gatk.broadinstitute.org/hc/en-us/articles/360036194592-Getting-started-with-GATK4)

    GATK (pronounced "Gee-ay-tee-kay", not "Gat-kay"), stands for GenomeAnalysisToolkit. It is a collection of command-line tools for analyzing high-throughput sequencing data with a primary focus on variant discovery. The tools can be used individually or chained together into complete workflows. We provide end-to-end workflows, called GATK Best Practices, tailored for specific use cases.  
    
- [pyvcf](https://pyvcf.readthedocs.io/en/latest/)

- [HTseq](https://htseq.readthedocs.io/en/master/)

    HTSeq is a Python package that provides infrastructure to process data from high-throughput sequencing assays. HTSeq is currently developed by Fabio Zanini at University of New South Wales in Sydney (fabio dot zanini at unsw dot edu dot au) and Simon Anders (anders at embl dot de) at EMBL Heidelberg (Genome Biology Unit). HTseq supports FASTA, FASTQ, SAM, VCF, GFF, and BED file formats. 

    [HTseq documentation](https://htseq.readthedocs.io/en/master/)


    HTseq documentation will teach you how to get started. It explains how to install it, and then walks you through analysis steps with explicit examples. 


### online sources

- books
    https://hub.packtpub.com/processing-next-generation-sequencing-datasets-using-python/

- videos:
    - [](https://youtu.be/j4qpJ8sVjT0)



## Population Genetics


- Libraries
- Numpy

- Biopython
- Bio.PopGen

Biopython makes Population Genetics functionality available in the Bio.PopGen module.

[Bio.PopGen documentation](https://biopython.org/wiki/PopGen)

- pygenomics
  
To see details about pygenomics use [pygenomics repository](https://github.com/tiagoantao/pygenomics)

- pypop


PyPop (Python for Population Genomics) is an environment developed by the Thomson lab for doing large-scale population genetic analyses including: (1) conformity to Hardy-Weinberg expectations, (2) tests for balancing or directional selection; (3) estimates of haplotype frequencies (and their distributions) and measures and tests of significance for linkage disequilibrium (LD). 
[PyPop Documentation](http://pypop.org/)



## Phylogenetics

- Biopython
- Bio.Phylo


This module provides classes, functions and I/O support for working with phylogenetic trees.
[Bio.Phylo documentation](https://biopython.org/wiki/Phylo)

- DendroPy

A Python library for phylogenetics and phylogenetic computing: reading, writing, simulation, processing and manipulation of phylogenetic trees (phylogenies) and characters.
[DendroPy documentation](https://pypi.org/project/DendroPy/)


Online Sources:


- Videos
    - Biopython Tutorial: Construct a phylogenetic tree
        - [](https://www.youtube.com/watch?v=8ICbCmZtuSk)



## Working with protein data bank

Libraries:

- BioPython
- Bio.PDB

[Bio.PDB documentation](https://biopython.org/docs/1.75/api/Bio.PDB.html)


Online Resources:


- Videos:
    - BioPython structure analysis
        - [](https://www.youtube.com/watch?v=mL8NPpRxgJA)


## Working with Big Data


