Python tools
===============

biopython
----------

The Biopython Project is an international association of developers of freely available Python (https://www.python.org) tools for computational molecular biology. The Biopython web site (http://www.biopython.org) provides an online resource for modules, scripts, and web links for developers of Python-based software for bioinformatics use and research. Basically, the goal of Biopython is to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes. Biopython features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.
 

[biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html)
    
    
    
    
    
    

Processing NGS data
----------

### online sources
#### https://hub.packtpub.com/processing-next-generation-sequencing-datasets-using-python/



#### HTseq

HTSeq is a Python package that provides infrastructure to process data from high-throughput sequencing assays. HTSeq is currently developed by Fabio Zanini at University of New South Wales in Sydney (fabio dot zanini at unsw dot edu dot au) and Simon Anders (anders at embl dot de) at EMBL Heidelberg (Genome Biology Unit). HTseq supports FASTA, FASTQ, SAM, VCF, GFF, and BED file formats. 


[HTseq documentation](https://htseq.readthedocs.io/en/master/)

HTseq documentation will teach you how to get started. It explains how to install it, and then walks you through analysis steps with explicit examples. 

note: HTSeq is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.




----------




